### ISSUE TITLE
ensuring text entry fields have appropriate aria labels for the page types listed below

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#925]
Similar to the issue we found for the Answer X of Y page (which is now fixed) we have other page types with text entry where there aren't currently labels or aria labels for those text fields. These include the following added as a check list so that we can tck them off when done rather than have a separate issue for each:

- [x] Dictation
- [x] Documentation
- [x] Gap Fill (in each mode)
- [x] Model Answer
- [x] Modify Exercise - Partial fix committed but not sure about the other two text boxes.
- [x] Dialogue
- [x] Opinion
- [ ] Drag and Drop labelling
- [x] Text correction

There might be others I have missed?

The next question is should the default label be the same for each of these or should it change for any of them? e.g. For Answer X of Y we have used "Answer" as the default which in turn becomes Answer 1, Answer 2, Answer 3 etc for each of the fields. However I'm just finishing the Model answer page and wondered if that should also be "Answer" by default or perhaps "Response"? Likewise what about the other page types?

Dictation =Answer?
Documentation = Answer?
Gap Fill = Answer?
Model Answer = Answer?
Modify Exercise = Answer? Partial fix committed but not sure about the other two text boxes.

Dialogue = Perhaps the aria-label should be the question otherwise a screen reader user isn't hearing the question. Also when they get it correct they hear "Tick" but should be a language string and perhaps default value should be "Correct"

**Comment by ronm123** [comment]
Hi all
The answer x of y page was listed as a separate issue that is done and closed now and I've committed the model answer solution and ticked that off above too. I'm going to work on the modify exercise next but I'm adding some notes and examples here of the strings and code used so far in case anyone else works on the other page types listed above. 

Example xwd entries:
Top XQ
`<answerFieldLabel label="Answer Field Hidden Aria-Label" type="TextInput" wysiwyg="false" language="true" mandatory="true" defaultValue="Answer" tooltip="This is a hidden label specifically for screen reader users so that each text field is labelled and spoken as Answer 1, Answer 2 etc."/>`

Modle Answer
`<answerFieldLabel label="Answer Field Hidden Aria-Label" type="TextInput" wysiwyg="false" language="true" mandatory="true" defaultValue="Answer" tooltip="This is a hidden label specifically for screen reader users so that the text field where they should respond is labelled and the label is spoken to the screenreader user. e.g. Answer."/>`

models
TopXQ
`var answerFieldLabel = x_currentPageXML.getAttribute("answerFieldLabel");
                if (answerFieldLabel === undefined | answerFieldLabel === null) {
                    answerFieldLabel = "Answer";
                }
 for (i = 0; i < amountOfAnswers; i++) {
                    var actualFieldLabel = answerFieldLabel + ' ' + (i+1);				
				 $optionHolder.append('<div class="answer"><input class="input-answer" id="input-answer-'+i+'" aria-label="'+actualFieldLabel+'"/><span id="topXQ-result-' + i + '"></span></div>');
                    j++;
                }	`

Model answer
`var answerFieldLabel = x_currentPageXML.getAttribute("answerFieldLabel");
			if (answerFieldLabel === undefined | answerFieldLabel === null) {
				answerFieldLabel = "Answer";
			}
			$("#answerTxt").attr({"aria-label":answerFieldLabel});`

**Comment by ronm123** [comment]
Hi all
on the modify exercise page we now have the following but not yet committed:

`<textarea id="initText" rows="5" aria-readonly="true" readonly></textarea>
	<p id="answer"></p>
	<textarea id="answerTxt" aria-label="Answer" rows="5"></textarea>
	<button id="button"></button>
	<textarea id="modelTxt" rows="5" readonly aria-live="polite"></textarea>`

The last text area doesn't appear until clicking the feedback button.

The Wave Accessibility checking tool was reporting missing form labels for each text area so I added aria-readonly="true" to the first field and aria-label="Answer" to the second. However Wave still reports missing labels for the first and last textareas, I can't find anything definitive to tell me if those textareas should also have aria-label="blah" or not. It would seem duplication to do that and I'm not sure if this is just false reporting by wave because neither lighthouse or axe report those particular issues. Anyone know the answer?

**Comment by ronm123** [comment]
BTW I should add that if we do need aria-labels on those read only text areas then it would mean at least 2 if not 3 additional language field options are needed in the wizard e.g. a label for each field or at least one for answer and another for the two read only fields. So I think not something we should do unless really needed.  @JohnSmith-LT @torinfo @FayCross any thoughts on this?

**Comment by JohnSmith-LT** [comment]
I'm not really up on all the accessibility guidance at the moment but is this where we could use aria-labelledby to reference another  element if that is more relevant?

**Comment by FayCross** [comment]
On the modify exercise page I'm not sure we necessarily need the read only text areas as text areas. I think they have probably only been done like this so they look visually like the text area that is editable. Would it make more sense to just have these as normal div / p elements and add some styling to make them look similar (border, margins etc.). I am happy to do this if you agree.

**Comment by JohnSmith-LT** [comment]
I did think the same but not sure whether click and Ctrl-A, Ctrl-C would be used much on this page and if so an input field limits Ctrl-A to it's content whereas a <div> doesn't... we could perhaps mitigate with some js to catch Ctrl-A... depends on how this page is actually used and how often I suppose...

**Comment by FayCross** [comment]
I don't think this would be a huge problem as the editable text field initially contains the same text as the first read only text field anyway. Perhaps that could do with a 'reset' button so you can easily return to the initial version

**Comment by ronm123** [comment]
I think a reset button would be good but I quite like the way that the 3 text areas appear just can’t find the answer to ether or not wave is reporting correctly that labels are also needed.

**Comment by ronm123** [comment]
Hi all
you will have seen that I've made quite a few commits resolving various accessibility issues and most of this has been issues reported by using the Chrome Wave plugin and a copy of our page types demonstrator. It would be really good to fix as many of these as possible to coincide with the public launch of our accessibility statement and a new release to go with that. I'm going to add a few more issues tagged under accessibility and resolve as many as I can but the list above of all the pages with missing form labels of some sort remains the biggest list of issues common to a number of pages. Would be good to get as many of these resolved as possible.

**Comment by ronm123** [comment]
@FayCross I think perhaps the drag and drop labelling was listed as a separate issue and fixed with that. I know it doesn't have text fields but I vaguely recall working with @JohnSmith-LT where we added the Tooltip description optional property. e.g. without  an optional property added to each target all a screen reader hears when tabbing to each target is drop zone for 1, drop zone for 2 etc which makes no sense. Adding that optional property the author could describe what each target covers so a screenreader user can decide where a label should be dropped. If an author wants to make their drag and drop labelling interaction accessible to screenreader users they have to add those optional properties and suitable descriptions. I'll close this issue.
